---
title: "Plot gene trackplots for cases where the mapped genes are not the nearest genes"
output:
  workflowr::wflow_html:
    code_folding: hide
    toc: true
editor_options:
  chunk_output_type: console
---

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(fig.align = "center")
```

Load packages and some functions used in this analysis
```{r load-packages, message=FALSE, warning=FALSE}
suppressMessages(library(data.table))
suppressMessages(library(tidyverse))
suppressMessages(library(GenomicRanges))
suppressMessages(library(AnnotationDbi)) # match gene ID to gene symbol
suppressMessages(library(org.Hs.eg.db)) # match gene ID to gene symbol
suppressMessages(library(GenomicInteractions)) # visualize HiC plots
suppressMessages(library(rtracklayer)) # loading bigwigs/bed files
suppressMessages(library(bigsnpr)) # loading genotype data from 1000Genomes for LD calculation
suppressMessages(library(Gviz)) # make track plots
library(mapgen)
```

Settings
```{r settings}
data_dir <- "/project2/xinhe/shared_data/aFib_gene_finemapping"
genomic_annots_file <- file.path(data_dir, "GWAS_gene_mapping_annots_hg19.gr.rds")
ABC_file <- file.path(data_dir, "/ABC/Nasser2021/tissues/heart_ventricle-ENCODE_ABC.tsv.gz")
H3K27ac_file <- "/project2/xinhe/shared_data/aFib_gene_finemapping/trackplots/H3K27ac_heart_concat.bed"
DHS_file <- "/project2/xinhe/shared_data/aFib_gene_finemapping/trackplots/FetalHeart_E083-DNase_hg19_cleaned_narrowPeak.bed.gz"

finemapping_file <- "/project2/xinhe/shared_data/aFib_gene_finemapping/pipeline/finemapping_latest_noAtrium_nonDA/AF_finemapping_result_torusprior_122loci.rds"
genemapping_res_dir <- "/project2/xinhe/shared_data/aFib_gene_finemapping/gene_mapping_result/aFib_gene_mapping_result_latest_noAtrium_nonDA_newSuSiE/"

outdir <- genemapping_res_dir
dir.create(outdir, showWarnings = F, recursive = T)
```

## Load results

Load finemapping results
```{r load-finemapping-res}
finemapstats <- readRDS(finemapping_file) %>% dplyr::rename(pip = susie_pip)
finemapstats.gr <- process_finemapping_sumstats(finemapstats, 
                                                        snp = 'snp', chr = 'chr', 
                                                        pos = 'pos', pip = 'pip', 
                                                        pval = 'pval', zscore = 'zscore', 
                                                        cs = 'CS', locus = 'locus',  
                                                        pip.thresh = 0)
```

Load annotations
```{r load-gene-annot, message=FALSE, warning=FALSE}
genomic.annots <- readRDS(genomic_annots_file)

gene.annots <- genomic.annots$genes

gene.locations <- as.data.frame(gene.annots)[, c("seqnames", "start", "end", "gene_name", "strand")]
gene.locations <- gene.locations %>% dplyr::rename(chr = seqnames) %>% mutate(tss = start(resize(gene.annots, width = 1)))

# PC-HiC
pcHiC.gr <- genomic.annots$HiC
pcHiC.gr <- pcHiC.gr[pcHiC.gr$gene_name %in% gene.annots$gene_name,] # restrict to protein coding genes

# ABC
ABC.gr  <- process_ABC(fread(ABC_file), ABC.thresh = 0.015, full.element = TRUE, flank = 0)
ABC.gr <- ABC.gr[ABC.gr$gene_name %in% gene.annots$gene_name, ] # restrict to protein coding genes
ABC.gr$promoter_chr <- seqnames(ABC.gr)
ABC.gr$promoter_start <- ABC.gr$TargetGeneTSS
ABC.gr$promoter_end <- ABC.gr$TargetGeneTSS
ABC.gr$score <- ABC.gr$ABC.Score * 100

# H3K27ac
H3K27ac_peaks <- rtracklayer::import(H3K27ac_file)

# Fetal DHS
DHS_peaks <- rtracklayer::import(DHS_file)
```

Load gene mapping results
```{r load-gene-mapping-result}
genemapping_res <- readRDS(file.path(genemapping_res_dir, "aFib_Finemapped_GeneMapped_activepromoter.OCRs_enhancerloop.ABC.pcHiC.nearby20kb_dist.50000_genemapping_res.rds"))
genemapping_res$fractional_PIP <- genemapping_res$pip * genemapping_res$frac_pip

m <- match(genemapping_res$snp, finemapstats$snp)
genemapping_res$pval <- finemapstats$pval[m]
genemapping_res$zscore <- finemapstats$zscore[m]
```

```{r snp-view}
# SNP view (fractional PIP > 0.02)
high.conf.snp.df <- fread(file.path(genemapping_res_dir, 'aFib_Finemapped_SNPView_fractionalPIP_0.02.csv'))
```

```{r gene-view}
# Gene view
gene.view.df <- fread(file.path(genemapping_res_dir, 'aFib_Finemapped_GenePIP_0.1.csv'))

cat(sum(gene.view.df$gene_pip >= 0.8), "genes with gene PIP >= 0.8 \n")
cat(sum(gene.view.df$gene_pip >= 0.5), "genes with gene PIP >= 0.5 \n")
```

```{r block-view}
# Block view
block.view.df <- fread(file.path(genemapping_res_dir, 'aFib_Finemapped_LDBlockView_Nielsen_V2.csv'))
```

## High PIP genes (gene PIP >= 0.8) that are not the nearest genes 

Find the nearest gene for the top SNP in each locus

```{r nearest-genes}
finemapstats.gr <- finemapstats.gr[order(abs(finemapstats.gr$zscore), finemapstats.gr$pval, decreasing = TRUE), ]
topsnps.gr <- finemapstats.gr[!duplicated(finemapstats.gr$locus), ]

nearest_tss_genes <- find_nearest_genes(topsnps.gr, gene.annots, 
                                        dist.to = "tss", 
                                        cols.to.keep = c('snp','chr','pos','locus','nearest_gene'))
nearest_genebody_genes <- find_nearest_genes(topsnps.gr, gene.annots, 
                                             dist.to = "genebody",
                                             cols.to.keep = c('snp','chr','pos','locus','nearest_gene'))

topsnps.gr$nearest_tss <- nearest_tss_genes$nearest_gene
topsnps.gr$nearest_genebody <- nearest_genebody_genes$nearest_gene

m <- match(block.view.df$locus, topsnps.gr$locus)
block.view.df$nearest_tss <- topsnps.gr$nearest_tss[m]
block.view.df$nearest_gene_body <- topsnps.gr$nearest_genebody[m]

block.view.df <- block.view.df %>% 
  dplyr::rename("Gene CS (80%)" = "80% Gene Credible Set") %>% 
  dplyr::select("locus", "Gene CS (80%)", "Top Genes", "Top Gene PIP", "nearest_gene_body", "nearest_tss")

high_pip_genes <- gene.view.df[gene.view.df$gene_pip >= 0.8, ]$gene_name
high_pip_locus_block_view.df <- block.view.df %>% filter(`Top Genes` %in% high_pip_genes)
```

High PIP genes (gene PIP >= 0.8) that are not the nearest genes by distance to gene body:
```{r high_pip_genes_not_nearest_genebody}
m <- NULL
for(i in 1:nrow(high_pip_locus_block_view.df)){
  locus_nearest_genes <- unlist(strsplit(high_pip_locus_block_view.df$nearest_gene_body[i], split=","))
  if(!high_pip_locus_block_view.df$`Top Genes`[i] %in% locus_nearest_genes){
    m <- c(m, i)
  }
}

high_pip_locus_block_view.df[m, c("locus", "Top Genes", "Top Gene PIP", "nearest_gene_body", "nearest_tss")]
```

High PIP genes (gene PIP >= 0.8) that are not the nearest genes by distance to TSS:
```{r high_pip_genes_not_nearest_tss}
m <- NULL
for(i in 1:nrow(high_pip_locus_block_view.df)){
  locus_nearest_genes <- unlist(strsplit(high_pip_locus_block_view.df$nearest_tss[i], split=","))
  if(!high_pip_locus_block_view.df$`Top Genes`[i] %in% locus_nearest_genes){
    m <- c(m, i)
  }
}

high_pip_locus_block_view.df[m, c("locus", "Top Genes", "Top Gene PIP", "nearest_tss", "nearest_gene_body")]
```

High PIP genes (gene PIP >= 0.8) that are not the nearest genes (by distance to TSS or gene body):
```{r high_pip_genes_not_nearest_genebody_tss}
m <- NULL
for(i in 1:nrow(high_pip_locus_block_view.df)){
  locus_nearest_genes <- unique(c( unlist(strsplit(high_pip_locus_block_view.df$nearest_gene_body[i], split=",")) ,
                                   unlist(strsplit(high_pip_locus_block_view.df$nearest_tss[i], split=",")) ))
  
  if(!high_pip_locus_block_view.df$`Top Genes`[i] %in% locus_nearest_genes){
    m <- c(m, i)
  }
}

high_pip_locus_block_view.df[m, c("locus", "Top Genes", "Top Gene PIP", "nearest_tss", "nearest_gene_body")]
```

Summary table for the high PIP genes (gene PIP >= 0.8) that are not the nearest genes by distance to TSS
```{r high_pip_genes_not_nearest_tss_summary_table}
m <- NULL
for(i in 1:nrow(high_pip_locus_block_view.df)){
  locus_nearest_genes <- unlist(strsplit(high_pip_locus_block_view.df$nearest_tss[i], split=","))
  if(!high_pip_locus_block_view.df$`Top Genes`[i] %in% locus_nearest_genes){
    m <- c(m, i)
  }
}

high_pip_genes_not_nearest_tss.df <- high_pip_locus_block_view.df[m, c("locus", "Top Genes", "Top Gene PIP", "Gene CS (80%)", "nearest_tss")]

high_pip_gene_evidence.df <- read.csv(file.path(outdir, 'gene_view_highPIP_0.8_evidence.csv'))

summary_table.df <- high_pip_genes_not_nearest_tss.df %>% 
  dplyr::rename("Gene" = "Top Genes", "Gene PIP" = "Top Gene PIP", "Nearest Gene" = "nearest_tss") %>%
  dplyr::left_join(high_pip_gene_evidence.df[, c("Gene", "Plausible.AF.risk.gene", "Molecular.function", "Literature.Reference")], by = "Gene") %>%
  dplyr::select("locus", "Gene", "Gene PIP", "Plausible.AF.risk.gene", "Molecular.function", "Literature.Reference", "Nearest Gene")

readr::write_csv(summary_table.df, file.path(outdir, 'high_pip_genes_not_nearest_tss_summary_table.csv'))
```

## Trackplots for the genes that are not the nearest genes

```{r prepare-trackplot-data, message=FALSE, warning=FALSE}
# Load the txdb object of the gene annotations
txdb <- AnnotationDbi::loadDb("/project2/xinhe/kevinluo/gene-level-finemapping/annot/gene_annotations/gencode.v19.annotation.gtf.sqlite")
big.snp <- bigsnpr::snp_attach('/project2/xinhe/1kg/bigsnpr/EUR_variable_1kg.rds')

atac_data.dir <- "/project2/xinhe/shared_data/aFib_gene_finemapping/trackplots/"
atac_data_files <- c(file.path(atac_data.dir, "Hg19_Cardiomyocyte-TileSize-500-normMethod-ReadsInTSS-ArchR.bw.bedGraph"),
                     file.path(atac_data.dir, "Hg19_Endothelial-TileSize-500-normMethod-ReadsInTSS-ArchR.bw.bedGraph"),
                     file.path(atac_data.dir, "Hg19_Fibroblast-TileSize-500-normMethod-ReadsInTSS-ArchR.bw.bedGraph"))

atac_data <- lapply(atac_data_files, function(x){rtracklayer::import(x)})
names(atac_data) <- c("CM", "Endo", "Fibro")

peaks <- list("H3K27ac" = H3K27ac_peaks, "Fetal DHS" = DHS_peaks)
HiC_loops <- list("PC-HiC" = pcHiC.gr, "ABC" = ABC.gr)
```

### FGF9

  * FGF9 (Fibroblast Growth Factor 9). Diseases associated with FGF9 include Multiple Synostoses Syndrome 3 and Multiple Synostoses Syndrome. Among its related pathways are Nervous system development and Negative regulation of FGFR1 signaling. Gene Ontology (GO) annotations related to this gene include growth factor activity and fibroblast growth factor receptor binding. An important paralog of this gene is FGF20.

  * Nearest gene: SGCG (Sarcoglycan Gamma). Diseases associated with SGCG include Muscular Dystrophy, Limb-Girdle, Autosomal Recessive 5 and Muscular Dystrophy. Among its related pathways are Acute viral myocarditis and DREAM Repression and Dynorphin Expression. An important paralog of this gene is SGCZ.

```{r FGF9-info}
genes_of_interest <- c("FGF9", "SGCG")
(genes.df <- gene.locations %>% filter(gene_name %in% genes_of_interest))

high.conf.snp.df[high.conf.snp.df$`Gene Linked` %in% genes_of_interest, c(1:3,6:13)] %>%
  arrange(-PIP) %>% as.data.frame()
```

```{r FGF9-trackplot, fig.width=12, fig.height=8, warning=FALSE}
region <- get_gene_region(genemapping_res, "FGF9", ext = 24000)
start(region) <- 22.2*1e6
end(region) <- 23.9*1e6

finemapping_annot_trackplot(finemapstats.gr,
                            region,
                            gene.annots,
                            bigSNP = big.snp,
                            genome = "hg19",
                            genetrack_db = "txdb",
                            txdb = txdb,
                            filter_protein_coding_genes = TRUE,
                            countsdata = atac_data,
                            data_colors = c("red", "green", "purple"),
                            data_ylim = c(0,0.8),
                            peaks = peaks,
                            HiC_loops = HiC_loops,
                            filter_HiCloops_genes = genes_of_interest,
                            highlight_snps = "topSNP",
                            verbose = FALSE)

```

### CALU

  * CALU (Calumenin). Diseases associated with CALU include Coumarin Resistance. Among its related pathways are Regulation of Insulin-like Growth Factor (IGF) transport and uptake by Insulin-like Growth Factor Binding Proteins (IGFBPs) and Warfarin Pathway, Pharmacodynamics. Gene Ontology (GO) annotations related to this gene include calcium ion binding.

  * Nearest gene: OPN1SW (Opsin 1, Short Wave Sensitive) is a Protein Coding gene. Diseases associated with OPN1SW include Tritanopia and Color Blindness. Among its related pathways are GPCR downstream signalling and Visual phototransduction. Gene Ontology (GO) annotations related to this gene include G protein-coupled receptor activity and photoreceptor activity. An important paralog of this gene is RHO.

```{r CALU-info}
genes_of_interest <- c("CALU", "OPN1SW")
(genes.df <- gene.locations %>% filter(gene_name %in% genes_of_interest))

high.conf.snp.df[high.conf.snp.df$`Gene Linked` %in% genes_of_interest, c(1:3,6:13)] %>% arrange(desc('Fractional PIP')) %>% as.data.frame()
```

```{r CALU-trackplot, fig.width=12, fig.height=8, warning=FALSE}
region <- get_gene_region(genemapping_res, "CALU", ext = 10000)
start(region) <- 128.375*1e6
end(region) <- 128.425*1e6

finemapping_annot_trackplot(finemapstats.gr,
                            region,
                            gene.annots,
                            bigSNP = big.snp,
                            genome = "hg19",
                            genetrack_db = "txdb",
                            txdb = txdb,
                            filter_protein_coding_genes = TRUE,
                            countsdata = atac_data,
                            data_colors = c("red", "green", "purple"),
                            data_ylim = c(0,0.8),
                            peaks = peaks,
                            HiC_loops = HiC_loops,
                            filter_HiCloops_genes = genes_of_interest,
                            highlight_snps = "topSNP",
                            verbose = FALSE)

```

### HCN4

  * HCN4 (Hyperpolarization Activated Cyclic Nucleotide Gated Potassium Channel 4). Diseases associated with HCN4 include Sick Sinus Syndrome 2 and Brugada Syndrome 8. Among its related pathways are Potassium Channels and Antiarrhythmic Pathway, Pharmacodynamics. Gene Ontology (GO) annotations related to this gene include identical protein binding and voltage-gated potassium channel activity. An important paralog of this gene is HCN2.

  * Nearest gene: NEO1 (Neogenin 1) is a Protein Coding gene. Diseases associated with NEO1 include Corneal Intraepithelial Neoplasm and Hemochromatosis Type 2. Among its related pathways are Angiogenesis (CST) and Netrin-1 signaling. Gene Ontology (GO) annotations related to this gene include signaling receptor activity and cadherin binding. An important paralog of this gene is DCC.

```{r HCN4-info}
genes_of_interest <- c("HCN4", "NEO1")
(genes.df <- gene.locations %>% filter(gene_name %in% genes_of_interest))

topsnps.gr[topsnps.gr$locus == c(1413,1414),]

high.conf.snp.df[high.conf.snp.df$`Gene Linked` %in% genes_of_interest, c(1:3,6:13)] %>% arrange(desc('Fractional PIP')) %>% as.data.frame()
```

```{r HCN4-trackplot, fig.width=12, fig.height=8, warning=FALSE}
region <- get_gene_region(genemapping_res, "HCN4", ext = 20000)
start(region) <- 73327990

finemapping_annot_trackplot(finemapstats.gr,
                            region,
                            gene.annots,
                            bigSNP = big.snp,
                            genome = "hg19",
                            genetrack_db = "txdb",
                            txdb = txdb,
                            filter_protein_coding_genes = TRUE,
                            countsdata = atac_data,
                            data_colors = c("red", "green", "purple"),
                            data_ylim = c(0,0.8),
                            peaks = peaks,
                            HiC_loops = HiC_loops,
                            filter_HiCloops_genes = genes_of_interest,
                            highlight_snps = c("rs1979409", "rs7172038"),
                            highlight_colors = c("pink", "yellow"),
                            verbose = FALSE)

```

### BEND5
About the genes:

  * BEND5 (BEN Domain Containing 5) is also the nearest gene by TSS distance. Diseases associated with BEND5 include Leukoencephalopathy, Cystic, Without Megalencephaly. Predicted to enable DNA binding activity. Involved in negative regulation of transcription, DNA-templated. Predicted to be located in Golgi apparatus.

  * Nearest gene by gene body distance: AGBL4 (AGBL Carboxypeptidase 4). Diseases associated with AGBL4 include Congenital Mesoblastic Nephroma and Hemolytic Uremic Syndrome, Atypical 1. Among its related pathways are Cooperation of Prefoldin and TriC/CCT in actin and tubulin folding and Metabolism of proteins. Gene Ontology (GO) annotations related to this gene include tubulin binding and metallocarboxypeptidase activity. An important paralog of this gene is AGTPBP1.

```{r BEND5-info}
genes_of_interest <- c("BEND5", "AGBL4")
(genes.df <- gene.locations %>% filter(gene_name %in% genes_of_interest))

high.conf.snp.df[high.conf.snp.df$`Gene Linked` %in% genes_of_interest, c(1:3,6:13)] %>% arrange(desc('Fractional PIP')) %>% as.data.frame()
```

```{r BEND5-trackplot, fig.width=12, fig.height=8, message=FALSE, warning=FALSE}
region <- get_gene_region(genemapping_res, "BEND5", ext = 10000)
start(region) <- 49*1e6
end(region) <- 51*1e6
# region

finemapping_annot_trackplot(finemapstats.gr,
                            region,
                            gene.annots,
                            bigSNP = big.snp,
                            genome = "hg19",
                            genetrack_db = "txdb",
                            txdb = txdb,
                            filter_protein_coding_genes = TRUE,
                            countsdata = atac_data,
                            data_colors = c("red", "green", "purple"),
                            data_ylim = c(0,0.8),
                            peaks = peaks,
                            HiC_loops = HiC_loops,
                            filter_HiCloops_genes = genes_of_interest,
                            highlight_snps = "topSNP",
                            genelabel_side = "above",
                            verbose = FALSE)
```

### PLN

  * PLN (Phospholamban), it is also the nearest gene by TSS distance. Diseases associated with PLN include Cardiomyopathy, Dilated, 1P and Cardiomyopathy, Familial Hypertrophic, 18. Among its related pathways are Beta-2 adrenergic-dependent CFTR expression and Antiarrhythmic Pathway, Pharmacodynamics. Gene Ontology (GO) annotations related to this gene include identical protein binding and enzyme inhibitor activity.

  * Nearest gene by gene body distance: SLC35F1: Predicted to enable transmembrane transporter activity. Predicted to be involved in transmembrane transport.

```{r PLN-info}
genes_of_interest <- c("PLN", "SLC35F1")
(genes.df <- gene.locations %>% filter(gene_name %in% genes_of_interest))

high.conf.snp.df[high.conf.snp.df$`Gene Linked` %in% genes_of_interest, c(1:3,6:13)] %>% arrange(desc('Fractional PIP')) %>% as.data.frame()
```

```{r PLN-trackplot, fig.width=12, fig.height=8, warning=FALSE}
region <- get_gene_region(genemapping_res, "PLN", ext = 10000)
start(region) <- 118.2*1e6
end(region) <- 118.9*1e6

finemapping_annot_trackplot(finemapstats.gr,
                            region,
                            gene.annots,
                            bigSNP = big.snp,
                            genome = "hg19",
                            genetrack_db = "txdb",
                            txdb = txdb,
                            filter_protein_coding_genes = TRUE,
                            countsdata = atac_data,
                            data_colors = c("red", "green", "purple"),
                            data_ylim = c(0,0.8),
                            peaks = peaks,
                            HiC_loops = HiC_loops,
                            filter_HiCloops_genes = genes_of_interest,
                            highlight_snps = "topSNP",
                            verbose = FALSE)
```

### TAB2

  * TAB2 (TGF-Beta Activated Kinase 1 (MAP3K7) Binding Protein 2), is also the nearest gene by TSS distance. Diseases associated with TAB2 include Congenital Heart Defects, Multiple Types, 2 and Polyvalvular Heart Disease Syndrome. Among its related pathways are SARS-CoV-2 Infection and TNF Superfamily - Human Ligand-Receptor Interactions and their Associated Functions. Gene Ontology (GO) annotations related to this gene include K63-linked polyubiquitin modification-dependent protein binding. An important paralog of this gene is TAB3.

  * Nearest gene by gene body distance: UST (Uronyl 2-Sulfotransferase). Diseases associated with UST include Breast Pericanalicular Fibroadenoma and Acute Retinal Necrosis Syndrome. Among its related pathways are Chondroitin sulfate/dermatan sulfate metabolism and Metabolism. Gene Ontology (GO) annotations related to this gene include sulfotransferase activity and galactosylceramide sulfotransferase activity. An important paralog of this gene is HS2ST1.

```{r TAB2-info}
genes_of_interest <- c("TAB2", "UST")
(genes.df <- gene.locations %>% filter(gene_name %in% genes_of_interest))

high.conf.snp.df[high.conf.snp.df$`Gene Linked` %in% genes_of_interest, c(1:3,6:13)] %>% arrange(desc('Fractional PIP')) %>% as.data.frame()
```

```{r TAB2-trackplot, fig.width=12, fig.height=8, warning=FALSE}
region <- get_gene_region(genemapping_res, "TAB2", ext = 10000)
start(region) <- 149*1e6
end(region) <- 149.8*1e6
# region

finemapping_annot_trackplot(finemapstats.gr,
                            region,
                            gene.annots,
                            bigSNP = big.snp,
                            genome = "hg19",
                            genetrack_db = "txdb",
                            txdb = txdb,
                            filter_protein_coding_genes = TRUE,
                            countsdata = atac_data,
                            data_colors = c("red", "green", "purple"),
                            data_ylim = c(0,0.8),
                            peaks = peaks,
                            HiC_loops = HiC_loops,
                            filter_HiCloops_genes = genes_of_interest,
                            highlight_snps = "topSNP",
                            verbose = FALSE)
```

### ETV1

  * ETV1 (ETS Variant Transcription Factor 1), is also the nearest gene by TSS distance. Diseases associated with ETV1 include Ewing Sarcoma and Gastrointestinal Stromal Tumor. Among its related pathways are P38 MAPK Signaling Pathway (sino) and MAPK-Erk Pathway. Gene Ontology (GO) annotations related to this gene include DNA-binding transcription factor activity and RNA polymerase II cis-regulatory region sequence-specific DNA binding. An important paralog of this gene is ETV5.

  * Nearest gene by gene body distance: DGKB (Diacylglycerol Kinase Beta) is a Protein Coding gene. Diseases associated with DGKB include Hermansky-Pudlak Syndrome 1. Among its related pathways are GPCR downstream signalling and Response to elevated platelet cytosolic Ca2+. Gene Ontology (GO) annotations related to this gene include calcium ion binding and diacylglycerol kinase activity. An important paralog of this gene is DGKG.

```{r ETV1-info}
genes_of_interest <- c("ETV1", "DGKB")
(genes.df <- gene.locations %>% filter(gene_name %in% genes_of_interest))

high.conf.snp.df[high.conf.snp.df$`Gene Linked` %in% genes_of_interest, c(1:3,6:13)] %>% arrange(desc('Fractional PIP')) %>% as.data.frame()
```

```{r ETV1-trackplot, fig.width=12, fig.height=8, warning=FALSE}
region <- get_gene_region(genemapping_res, "ETV1", ext = 10000)
start(region) <- 13.9*1e6
end(region) <- 15.0*1e6
# region

finemapping_annot_trackplot(finemapstats.gr,
                            region,
                            gene.annots,
                            bigSNP = big.snp,
                            genome = "hg19",
                            genetrack_db = "txdb",
                            txdb = txdb,
                            filter_protein_coding_genes = TRUE,
                            countsdata = atac_data,
                            data_colors = c("red", "green", "purple"),
                            data_ylim = c(0,0.8),
                            peaks = peaks,
                            HiC_loops = HiC_loops,
                            filter_HiCloops_genes = genes_of_interest,
                            highlight_snps = "topSNP",
                            verbose = FALSE)
```

### DBX1

  * DBX1 (Developing Brain Homeobox 1) is also the nearest gene by TSS distance. Diseases associated with DBX1 include Drug Psychosis and Drug-Induced Mental Disorder. Gene Ontology (GO) annotations related to this gene include sequence-specific DNA binding. An important paralog of this gene is DBX2.

  * Nearest gene by gene body distance: NAV2 (Neuron Navigator 2) is a Protein Coding gene. Diseases associated with NAV2 include Neuroblastoma 1 and Neuroblastoma. Gene Ontology (GO) annotations related to this gene include heparin binding and helicase activity. An important paralog of this gene is NAV3.

```{r DBX1-info}
genes_of_interest <- c("DBX1", "NAV2")
(genes.df <- gene.locations %>% filter(gene_name %in% genes_of_interest))

high.conf.snp.df[high.conf.snp.df$`Gene Linked` %in% genes_of_interest, c(1:3,6:13)] %>% arrange(desc('Fractional PIP')) %>% as.data.frame()
```

```{r DBX1-trackplot, fig.width=12, fig.height=8, warning=FALSE}
region <- get_gene_region(genemapping_res, "DBX1", ext = 10000)
start(region) <- 19.3*1e6
end(region) <- 20.2*1e6
# region

finemapping_annot_trackplot(finemapstats.gr,
                            region,
                            gene.annots,
                            bigSNP = big.snp,
                            genome = "hg19",
                            genetrack_db = "txdb",
                            txdb = txdb,
                            filter_protein_coding_genes = TRUE,
                            countsdata = atac_data,
                            data_colors = c("red", "green", "purple"),
                            data_ylim = c(0,0.8),
                            peaks = peaks,
                            HiC_loops = HiC_loops,
                            filter_HiCloops_genes = genes_of_interest,
                            highlight_snps = "topSNP",
                            verbose = FALSE)
```


### KCNN3

  * KCNN3 (Potassium Calcium-Activated Channel Subfamily N Member 3) is also the nearest gene by gene body distance. Diseases associated with KCNN3 include Zimmermann-Laband Syndrome 3 and Zimmermann-Laband Syndrome. Among its related pathways are Potassium Channels and Transmission across Chemical Synapses. Gene Ontology (GO) annotations related to this gene include protein heterodimerization activity and calcium-activated potassium channel activity. An important paralog of this gene is KCNN2.

  * Nearest gene by TSS distance: ADAR (Adenosine Deaminase RNA Specific). Diseases associated with ADAR include Dyschromatosis Symmetrica Hereditaria and Aicardi-Goutieres Syndrome 6. Among its related pathways are Pathways of nucleic acid metabolism and innate immune sensing and mRNA Editing: A to I Conversion. Gene Ontology (GO) annotations related to this gene include RNA binding and adenosine deaminase activity. An important paralog of this gene is ADARB2.

```{r KCNN3-info}
genes_of_interest <- c("KCNN3", "ADAR")
(genes.df <- gene.locations %>% filter(gene_name %in% genes_of_interest))

topsnps.gr[topsnps.gr$locus == c(77,78),]

high.conf.snp.df[high.conf.snp.df$`Gene Linked` %in% genes_of_interest, c(1:3,6:13)] %>% arrange(desc('Fractional PIP')) %>% as.data.frame()
```


```{r KCNN3-trackplot, fig.width=12, fig.height=8, warning=FALSE}
region <- get_gene_region(genemapping_res, "KCNN3", ext = 10000)
gene.locations[gene.locations$gene_name == "ADAR", ]
gene.locations[gene.locations$gene_name == "KCNN3", ]

start(region) <- 154554000
end(region) <- 154870000

finemapping_annot_trackplot(finemapstats.gr,
                            region,
                            gene.annots,
                            bigSNP = big.snp,
                            genome = "hg19",
                            genetrack_db = "txdb",
                            txdb = txdb,
                            filter_protein_coding_genes = TRUE,
                            countsdata = atac_data,
                            data_colors = c("red", "green", "purple"),
                            data_ylim = c(0,0.8),
                            peaks = peaks,
                            HiC_loops = HiC_loops,
                            filter_HiCloops_genes = genes_of_interest,
                            highlight_snps = c("rs4999127", "rs11264280"),
                            highlight_colors = c("pink", "yellow"),
                            genelabel_side = "above",
                            verbose = FALSE)

```


### TTN

  * TTN (Titin) is also the nearest gene by gene body distance. Diseases associated with TTN include Myopathy, Myofibrillar, 9, With Early Respiratory Failure and Salih Myopathy. Among its related pathways are Response to elevated platelet cytosolic Ca2+ and Striated muscle contraction pathway. Gene Ontology (GO) annotations related to this gene include nucleic acid binding and identical protein binding.

  * Nearest gene by TSS distance: PLEKHA3 (Pleckstrin Homology Domain Containing A3) is a Protein Coding gene. Diseases associated with PLEKHA3 include Rhizomelic Chondrodysplasia Punctata, Type 3 and Myasthenic Syndrome, Congenital, 10. Among its related pathways are Cholesterol and Sphingolipids transport / Transport from Golgi and ER to the apical membrane (normal and CF) and PI Metabolism. Gene Ontology (GO) annotations related to this gene include phospholipid binding and 1-phosphatidylinositol binding. An important paralog of this gene is PLEKHA8.

```{r TTN-info}
genes_of_interest <- c("TTN", "PLEKHA3")
(genes.df <- gene.locations %>% filter(gene_name %in% genes_of_interest))

high.conf.snp.df[high.conf.snp.df$`Gene Linked` %in% genes_of_interest, c(1:3,6:13)] %>% arrange(desc('Fractional PIP')) %>% as.data.frame()
```


```{r TTN-trackplot, fig.width=12, fig.height=8, warning=FALSE}
region <- get_gene_region(genemapping_res, "TTN", ext = 10000)
start(region) <- 179301000
end(region) <- 179690000

finemapping_annot_trackplot(finemapstats.gr,
                            region,
                            gene.annots,
                            bigSNP = big.snp,
                            genome = "hg19",
                            genetrack_db = "txdb",
                            txdb = txdb,
                            filter_protein_coding_genes = TRUE,
                            countsdata = atac_data,
                            data_colors = c("red", "green", "purple"),
                            data_ylim = c(0,0.8),
                            peaks = peaks,
                            HiC_loops = HiC_loops,
                            filter_HiCloops_genes = genes_of_interest,
                            highlight_snps = "topSNP",
                            genelabel_side = "above",
                            verbose = FALSE)

```

### ATXN1

  * ATXN1 (Ataxin 1) is also the nearest gene by gene body distance. Diseases associated with ATXN1 include Spinocerebellar Ataxia 1 and Primary Cerebellar Degeneration. Among its related pathways are Regulation of activated PAK-2p34 by proteasome mediated degradation and Akt Signaling. Gene Ontology (GO) annotations related to this gene include identical protein binding and chromatin binding. An important paralog of this gene is ATXN1L.

  * Nearest gene by TSS distance: GMPR (Guanosine Monophosphate Reductase). Diseases associated with GMPR include Viral Gastritis and Cystathioninuria. Among its related pathways are Metabolism and Nucleotide salvage. Gene Ontology (GO) annotations related to this gene include GMP reductase activity. An important paralog of this gene is GMPR2.

```{r ATXN1-info}
genes_of_interest <- c("ATXN1", "GMPR")
(genes.df <- gene.locations %>% filter(gene_name %in% genes_of_interest))

high.conf.snp.df[high.conf.snp.df$`Gene Linked` %in% genes_of_interest, c(1:3,6:13)] %>% arrange(desc('Fractional PIP')) %>% as.data.frame()
```

```{r ATXN1-trackplot, fig.width=12, fig.height=8, warning=FALSE}
region <- get_gene_region(genemapping_res, "ATXN1", ext = 10000)
start(region) <- 16200000
end(region) <- 16800000

finemapping_annot_trackplot(finemapstats.gr,
                            region,
                            gene.annots,
                            bigSNP = big.snp,
                            genome = "hg19",
                            genetrack_db = "txdb",
                            txdb = txdb,
                            filter_protein_coding_genes = TRUE,
                            countsdata = atac_data,
                            data_colors = c("red", "green", "purple"),
                            data_ylim = c(0,0.8),
                            peaks = peaks,
                            HiC_loops = HiC_loops,
                            filter_HiCloops_genes = genes_of_interest,
                            highlight_snps = "topSNP",
                            genelabel_side = "above",
                            verbose = FALSE)

```

### SSPN

  * SSPN (Sarcospan) is also the nearest gene by gene body distance. Diseases associated with SSPN include Autosomal Recessive Limb-Girdle Muscular Dystrophy Type 2F and Autosomal Recessive Limb-Girdle Muscular Dystrophy Type 2C. Among its related pathways are Clear cell renal cell carcinoma pathways and Muscular Dystrophies and Dystrophin-Glycoprotein Complex.

  * Nearest gene by TSS distance: BHLHE41 (Basic Helix-Loop-Helix Family Member E41). Diseases associated with BHLHE41 include Short Sleep, Familial Natural, 1 and Advanced Sleep Phase Syndrome, Familial, 1. Among its related pathways are Circadian Clock and Clear cell renal cell carcinoma pathways. Gene Ontology (GO) annotations related to this gene include protein homodimerization activity and RNA polymerase II cis-regulatory region sequence-specific DNA binding. An important paralog of this gene is BHLHE40.

```{r SSPN-info}
genes_of_interest <- c("SSPN", "BHLHE41")
(genes.df <- gene.locations %>% filter(gene_name %in% genes_of_interest))

high.conf.snp.df[high.conf.snp.df$`Gene Linked` %in% genes_of_interest, c(1:3,6:13)] %>% arrange(desc('Fractional PIP')) %>% as.data.frame()
```


```{r SSPN-trackplot, fig.width=12, fig.height=8, warning=FALSE}
region <- get_gene_region(genemapping_res, "SSPN", ext = 25000)
start(region) <- 26.25*1e6
end(region) <- 26.48*1e6
# region

finemapping_annot_trackplot(finemapstats.gr,
                            region,
                            gene.annots,
                            bigSNP = big.snp,
                            genome = "hg19",
                            genetrack_db = "txdb",
                            txdb = txdb,
                            filter_protein_coding_genes = TRUE,
                            countsdata = atac_data,
                            data_colors = c("red", "green", "purple"),
                            data_ylim = c(0,0.8),
                            peaks = peaks,
                            HiC_loops = HiC_loops,
                            filter_HiCloops_genes = genes_of_interest,
                            highlight_snps = "topSNP",
                            verbose = FALSE)

```

### PKP2

  * PKP2 is also the nearest gene by gene body distance. This gene encodes a member of the arm-repeat (armadillo) and plakophilin gene families. Plakophilin proteins contain numerous armadillo repeats, localize to cell desmosomes and nuclei, and participate in linking cadherins to intermediate filaments in the cytoskeleton. This gene may regulate the signaling activity of beta-catenin and is required to maintain transcription of genes that control intracellular calcium cycling including ryanodine receptor 2, ankyrin-B, triadin, and calcium channel, voltage-dependent, L type, alpha 1C. Mutations in this gene are associated with different inherited cardiac conditions including Arrythmogenic Cardiomyopathy, Brugada Syndrome, and Idiopathic Ventricular Fibrillation.

  * Nearest gene by TSS distance: YARS2. This gene encodes a mitochondrial protein that catalyzes the attachment of tyrosine to tRNA(Tyr). Mutations in this gene are associated with myopathy with lactic acidosis and sideroblastic anemia type 2 (MLASA2).

```{r PKP2-info}
genes_of_interest <- c("PKP2", "YARS2")
(genes.df <- gene.locations %>% filter(gene_name %in% genes_of_interest))

high.conf.snp.df[high.conf.snp.df$`Gene Linked` %in% genes_of_interest, c(1:3,6:13)] %>% arrange(desc('Fractional PIP')) %>% as.data.frame()
```

```{r PKP2-trackplot, fig.width=12, fig.height=8, message=FALSE, warning=FALSE}
region <- get_gene_region(genemapping_res, "PKP2", ext = 10000)
start(region) <- 32.87*1e6
end(region) <- 33.05*1e6
# region

finemapping_annot_trackplot(finemapstats.gr,
                            region,
                            gene.annots,
                            bigSNP = big.snp,
                            genome = "hg19",
                            genetrack_db = "txdb",
                            txdb = txdb,
                            filter_protein_coding_genes = TRUE,
                            countsdata = atac_data,
                            data_colors = c("red", "green", "purple"),
                            data_ylim = c(0,0.8),
                            peaks = peaks,
                            HiC_loops = HiC_loops,
                            filter_HiCloops_genes = genes_of_interest,
                            highlight_snps = "topSNP",
                            genelabel_side = "above",
                            verbose = FALSE)
```





